---
title: "Diabetes Kidney Tumor Gene Expressions ML"
author: "Janis Corona"
date: "12/23/2019"
output:
  word_document: default
  html_document:
    df_print: paged
  pdf_document: default
---

### This R-Markdown file uses R to analyze kidney tumor tissue gene expression profiles of Diabetes patients in Advanced, Early, or healthy states, performs predictive analytics on those genes, text mining, and compares Diabetic kidney disease to renal disease kidney tumor tissue from a previous study.

### GSE142025 from https://www.ncbi.nlm.nih.gov/geo

Overall design of the GSE142025 study is using kidney tumor and healthy kidney samples to analyze stages of diabetes mellitis. A total of 28 patients with biopsy-proven DN hospitalized from January 2015 to December 2016 in Shanghai Jiao Tong University Affiliated Sixth Peopleâ€™s Hospital were enrolled in the study. Nine control human kidney samples were obtained from the unaffected portion of tumor nephrectomies. RNA-seq was performed on 28 DN and 9 control samples.

Later on, machine learning with KNN and Random Forest is used to test out these samples on identifying the disease state derived, and also using a separate kidney tumor study that analyzed kidney tumors from healthy and renal disease patients. The actual data wasn't used, just the gene targets identified as the top 20 up regulated and top 20 increased fold change as well as top 20 down regulated genes from that study that were available in this data set to use machine learning on.


#### Read in the three data sets for Advanced, early, and control kidney tumor samples

```{r}
Advanced <- read.csv('Advanced_Diabetes.csv',
                          sep=',', header=TRUE,
                          na.strings=c('',' ','NA'))
head(Advanced)

```

```{r}

tail(Advanced)

```

```{r}
Early <- read.csv('Early_Diabetes.csv',
                          sep=',', header=TRUE,
                          na.strings=c('',' ','NA'))
head(Early)

```

```{r}

tail(Early)

```

```{r}
Control <- read.csv('Control_Group.csv',
                          sep=',', header=TRUE,
                          na.strings=c('',' ','NA'))
head(Control)

```

```{r}

tail(Control)

```



```{r, error=FALSE, message=FALSE, warning=FALSE}
library(dplyr)
```


## Assign no duplicate instances of genes
```{r}
Advanced1 <- Advanced[!duplicated(Advanced$Symbol),]
```

## Check that all the genes have 1 count each, they do
```{r}
AdvancedCounts <- Advanced1 %>% group_by(Symbol) %>% 
              summarise(counts = n())
dim(AdvancedCounts)
unique(AdvancedCounts$counts)

```

```{r}
Early1 <- Early[!duplicated(Early$Symbol),]
```

## Check that all the genes have 1 count each, they do
```{r}
EarlyCounts <- Early1 %>% group_by(Symbol) %>% 
              summarise(counts = n())
dim(EarlyCounts)
unique(EarlyCounts$counts)

```

```{r}
Control1 <- Control[!duplicated(Control$Symbol),]
```

## Check that all the genes have 1 count each, they do
```{r}
ControlCounts <- Control1 %>% group_by(Symbol) %>% 
              summarise(counts = n())
dim(ControlCounts)
unique(ControlCounts$counts)

```


#### Attach a field to each of the data sets of the row means for those samples unique gene
```{r}
row.names(Advanced1) <- Advanced1$Symbol
Advanced1 <- Advanced1[,2:22]
Advanced1$Gene_Means <- rowMeans(Advanced1)
```


```{r}
row.names(Early1) <- Early1$Symbol
Early1 <- Early1[,2:7]
Early1$Gene_Means <- round(rowMeans(Early1),3)
Advanced1$Gene_Means <- round(Advanced1$Gene_Means,3)
```


```{r}
row.names(Control1) <- Control1$Symbol
Control1 <- Control1[,2:10]
Control1$Gene_Means <- round(rowMeans(Control1),3)
```


```{r}
colnames(Advanced1)[22] <- "Advanced_DM_Means"
colnames(Early1)[7] <- "Early_DM_Means"
colnames(Control1)[10] <- "Control_noDM_Means"
```


```{r}
means <- cbind(Advanced1[22], Early1[7], Control1[10])
Combined <- cbind(means, Advanced1[1:21], Early1[1:6], Control1[1:9])
```

#### Create the fold change field to compare the change in early Diabetes Mellitis (DM) Kidney tumor gene expressions to healthy non DM kidney tissue gene expressions, Advanced DM to healthy, and Advanced to Early DM expressions.
```{r}
Fold_Change0 <- Combined %>% mutate(Fold_Change_Early_Control = Early_DM_Means/Control_noDM_Means)

Fold_Change1 <- Fold_Change0 %>% mutate(Fold_Change_Advanced_Control = Advanced_DM_Means/Control_noDM_Means)

Fold_Change2 <- Fold_Change1 %>% mutate(Fold_Change_Advanced_Early = Advanced_DM_Means/Early_DM_Means)

row.names(Fold_Change2) <- row.names(Combined)

Fold_Changes <- Fold_Change2[,c(40,41,42,1:39)]

```

#### Remove NaN's or Not a number and Inf when dividing by zero or a very small value
```{r}
Fold_Changes$Fold_Change_Early_Control <- gsub('NaN',0,Fold_Changes$Fold_Change_Early_Control)
Fold_Changes$Fold_Change_Early_Control <- gsub('Inf', 0, Fold_Changes$Fold_Change_Early_Control)
Fold_Changes$Fold_Change_Early_Control <- round(as.numeric(Fold_Changes$Fold_Change_Early_Control),3)

Fold_Changes$Fold_Change_Advanced_Control <- gsub('NaN',0,Fold_Changes$Fold_Change_Advanced_Control)
Fold_Changes$Fold_Change_Advanced_Control <- gsub('Inf', 0, Fold_Changes$Fold_Change_Advanced_Control)
Fold_Changes$Fold_Change_Advanced_Control <- round(as.numeric(Fold_Changes$Fold_Change_Advanced_Control),3)

Fold_Changes$Fold_Change_Advanced_Early <- gsub('NaN',0,Fold_Changes$Fold_Change_Advanced_Early)
Fold_Changes$Fold_Change_Advanced_Early <- gsub('Inf', 0, Fold_Changes$Fold_Change_Advanced_Early)
Fold_Changes$Fold_Change_Advanced_Early <- round(as.numeric(Fold_Changes$Fold_Change_Advanced_Early),3)

write.csv(Fold_Changes,'Fold_Changes.csv', row.names=TRUE)
```

```{r}
Top20_FC_Adv_Early <- Fold_Changes[order(Fold_Changes$Fold_Change_Advanced_Early, decreasing=TRUE)[0:20],]
write.csv(Top20_FC_Adv_Early,'Top20_FC_Adv-Early.csv', row.names=TRUE)
```


```{r}
Top20_FC_Adv_Control <- Fold_Changes[order(Fold_Changes$Fold_Change_Advanced_Control, decreasing=TRUE)[0:20],]
write.csv(Top20_FC_Adv_Control,'Top20_FC_Adv-Ctrl.csv',row.names=TRUE)
```


```{r}
Top20_FC_Early_Control <- Fold_Changes[order(Fold_Changes$Fold_Change_Early_Control, decreasing=TRUE)[0:20],]
write.csv(Top20_FC_Early_Control,'Top20_FC_Early-Ctrl.csv',row.names=TRUE)
```


#### Create the Differential Expression fields for Control-Early, Control-Advanced, and Early-Advanced so that negative differential expression values mean those genes are up regulated in the more diseased state compared to the healthy state.
```{r}
DE_Control_Adv <- Fold_Changes %>% mutate(DE_ctrl_adv = round(Control_noDM_Means-Advanced_DM_Means,3))

DE_Early_Adv <- DE_Control_Adv %>% mutate(DE_early_adv = round(Early_DM_Means-Advanced_DM_Means,3))

DE_Control_Early <- DE_Early_Adv %>% mutate(DE_ctrl_early = round(Control_noDM_Means-Early_DM_Means,3))

Differential <- DE_Control_Early[,c(43,45,44,1:42)]
row.names(Differential) <- row.names(Fold_Changes)

write.csv(Differential, 'Differential_FCs.csv', row.names=TRUE)
```

#### Since this is healthy - diseased, and less diseased-more diseased, positive values mean the diseased gene expression means are lower than the healthy gene expression levels or down regulated in diseased or more diseased. While the negative values mean those genes are up regulated in diseased or more diseased states compared to healthy or less diseased states.
```{r}
down_Ctrl_Adv <- Differential[order(Differential$DE_ctrl_adv, 
                                decreasing=TRUE),]
down_Ctrl_Early <- Differential[order(Differential$DE_ctrl_early, 
                                decreasing=TRUE),]
down_Early_Adv <- Differential[order(Differential$DE_early_adv, 
                                decreasing=TRUE),]

up_Ctrl_Adv <- Differential[order(Differential$DE_ctrl_adv, 
                                decreasing=FALSE),]
up_Ctrl_Early <- Differential[order(Differential$DE_ctrl_early, 
                                decreasing=FALSE),]
up_Early_Adv <- Differential[order(Differential$DE_early_adv, 
                                decreasing=FALSE),]

```

#### Top 20 downgraded genes
```{r}
Top20_down_Ctrl_Adv <- down_Ctrl_Adv[0:20,]
Top20_down_Ctrl_Early <- down_Ctrl_Early[0:20,]
Top20_down_Early_Adv <- down_Early_Adv[0:20,]

```

#### Top 20 upgraded genes
```{r}
Top20_up_Ctrl_Adv <- up_Ctrl_Adv[0:20,]
Top20_up_Ctrl_Early <- up_Ctrl_Early[0:20,]
Top20_up_Early_Adv <- up_Early_Adv[0:20,]

```

#### Write these up and down regulated genes to csv files

```{r}
write.csv(Top20_up_Ctrl_Adv, 'Top20_up_Ctrl_Adv.csv', row.names=TRUE)
write.csv(Top20_up_Ctrl_Early, 'Top20_up_Ctrl_Early.csv', row.names=TRUE)
write.csv(Top20_up_Early_Adv, 'Top20_up_Early_Adv.csv', row.names=TRUE)

write.csv(Top20_down_Ctrl_Adv, 'Top20_down_Ctrl_Adv.csv', row.names=TRUE)
write.csv(Top20_down_Ctrl_Early, 'Top20_down_Ctrl_Early.csv', row.names=TRUE)
write.csv(Top20_down_Early_Adv, 'Top20_down_Early_Adv.csv', row.names=TRUE)

```



### What are the top 20 genes up-regulated in Diabetes Mellitis Kidney disease of all three sets?

Control state versus the early DM state of up regulated genes
```{r}
Up1 <- as.data.frame(row.names(Top20_up_Ctrl_Early))
colnames(Up1) <- 'Gene'
Up1$dataType <- rep('early up regulated',20)
Up1 <- Up1[order(Up1$Gene),]

Up1
```

Early state versus the Advanced DM state of up regulated genes
```{r}
Up2 <- as.data.frame(row.names(Top20_up_Early_Adv))
colnames(Up2) <- 'Gene'
Up2$dataType <- rep('Advanced Early Up',20)
Up2 <- Up2[order(Up2$Gene),]
Up2

```

Control state versus the Advanced DM state of up regulated genes
```{r}
Up3 <- as.data.frame(row.names(Top20_up_Ctrl_Adv))
colnames(Up3) <- 'Gene'
Up3$dataType <- rep('Advanced Control Up', 20)
Up3 <- Up3[order(Up3$Gene),]
Up3
```

### What are the top 20 genes down-regulated in renal disease compared to healthy?

Control state versus the early DM state of down regulated genes
```{r}
down1 <- as.data.frame(row.names(Top20_down_Ctrl_Early))
colnames(down1) <- 'Gene'
down1$dataType <- rep('Early Control Down', 20)
down1 <- down1[order(down1$Gene),]
down1

```

Early state versus the Advanced DM state of down regulated genes
```{r}
down2 <- as.data.frame(row.names(Top20_down_Early_Adv))
colnames(down2) <- 'Gene'
down2$dataType <- rep('Advanced Early Down', 20)
down2 <- down2[order(down2$Gene),]
down2

```

Control state versus the Advanced DM state of down regulated genes
```{r}
down3 <- as.data.frame(row.names(Top20_down_Ctrl_Adv))
colnames(down3) <- 'Gene'
down3$dataType <- rep('Advanced Control Down',20)
down3 <- down3[order(down3$Gene),]
down3
```


### What are the top 20 genes that have the most fold change in the ratio of healthy to renal disease gene expression? Even the inverse fold change of disease to healthy would 
```{r}
FC <- as.data.frame(row.names(Top20_FC_Early_Control))
colnames(FC) <- 'Gene'
FC$dataType <- rep('Early Control Fold Change Up',20)
FC <- FC[order(FC$Gene),]
FC
```

```{r}
FC1 <- as.data.frame(row.names(Top20_FC_Adv_Early))
colnames(FC1) <- 'Gene'
FC1$dataType <- rep('Advanced Early Fold Change Up', 20)
FC1 <- FC1[order(FC1$Gene),]
FC1
```

```{r}
FC2 <- as.data.frame(row.names(Top20_FC_Adv_Control))
colnames(FC2) <- 'Gene'
FC2$dataType <- rep('Advanced Control Fold Change Up', 20)
FC2 <- FC2[order(FC2$Gene),]
FC2
```

```{r}
down <- rbind(down1,down2,down3)
down <- down[unique(down$Gene),]

up <- rbind(Up1, Up2, Up3)
up <- up[unique(up$Gene),]

fc <- rbind(FC, FC1, FC2)
fc <- fc[unique(fc$Gene),]

all <- rbind(down,up,fc)

write.csv(all, 'all_common_up_or_fc_and_down.csv', row.names=FALSE)
```

#### Common genes to most fold change and up regulated gene expressions are:
```{r}
head(all)
```

```{r}
tail(all)

```


### Now for some machine learning on predicting the dataType as one of these nine types of data sets
```{r}
unique(all$dataType)
```

### lets bring in the data from a kidney tumor study that was done on 12 kidney disease and 4 healthy kidney tissue samples for the top 20 down, up, and fold change increases of diseased to healthy comparisons. Note the gene expression values will differ, because this study used a different platform than the DM study in this Markdown file.
```{r}
renal_up20 <- read.csv('Up-regulated-20.csv', sep=',', header=TRUE,
                       na.strings=c('',' '))
colnames(renal_up20)[1] <- 'Gene'

renal_fc20 <- read.csv('Fold-Change-20.csv', sep=',', header=TRUE,
                       na.strings=c('',' '))
colnames(renal_fc20)[1] <- 'Gene'
renal_down20 <- read.csv('Down-regulated-20.csv', sep=',', header=TRUE,
                       na.strings=c('',' '))
colnames(renal_down20)[1] <- 'Gene'

```

What genes from the renal disease data were top genes for up, down, or increased fold change?
```{r}
renal_up20$dataType <- rep('Renal Disease Up regulated', 20)
renal_down20$dataType <- rep('Renal Disease Down regulated', 20)
renal_fc20$dataType <- rep('Renal Disease Increased Fold Change', 20)

# keep only the Gene field and the dataType fields
renal_up20 <- renal_up20[,c(1,22)]
renal_down20 <- renal_down20[,c(1,22)]
renal_fc20 <- renal_fc20[,c(1,21)]

combined_Renal <- rbind(renal_up20,renal_fc20,renal_down20)

Renal_up <- combined_Renal[0:40,]
Renal_down <- combined_Renal[41:60,]

```

Lets focus on those genes that were up regulated in both the DM and Renal Disease samples including fold change, and those in common for DM and Renal Disease samples that were down regulated.
```{r}
Renal_up$Regulation <- rep('increased', 40)
Renal_down$Regulation <- rep('decreased',20)

RENAL <- rbind(Renal_up,Renal_down)

RENAL <- RENAL[,c(1,3)]
RENAL$Disease <- rep('Renal Disease', length(RENAL$Gene))

down <- all[grep('Down',all$dataType),]
down$Regulation <- rep('decreased', length(down$Gene))
up <- all[grep('Up', all$dataType),]
up$Regulation <- rep('increased', length(up$Gene))

DIABETES <- rbind(down, up)
DIABETES <- DIABETES[,c(1,3)]
DIABETES$Disease <- rep('Diabetes Mellitis', length(DIABETES$Gene))

```

Use the unique genes in the DIABETES data set and the RENAL data set and combine
```{r}
diabetes <- DIABETES[unique(DIABETES$Gene),]
renal <- RENAL[unique(RENAL$Gene),]

Diabetes_renal_common_genes <- merge(diabetes,renal, by.x='Gene', by.y='Gene')
```

Only the XIST gene is common to both diseases for most fold change, up, and down regulation. This gene increased and was up regulated in both diabetes mellitis and renal disease tissue gene expressions.


#### Use only the diabetes data set with the unique genes to each of the types of disease states compared to each other and the healthy state.Get the gene expressions for those samples by merging this diabetes data set by Gene with the Differential data set
```{r}
Differential$Gene <- row.names(Differential)

expressions <- merge(diabetes, Differential, by.x='Gene', by.y='Gene')
Expressions <- expressions[unique(expressions$Gene),]
Expressions <- Expressions[complete.cases(Expressions$Fold_Change_Early_Control),]
Expressions <- Expressions[!duplicated(Expressions$Gene),]
```

Let us satisfy the curiosity of knowing how the genes that were top expressed in fold change, up regulation, or down regulation from the renal disease data to this data to see how the diabetes data does.
```{r}
curious <- merge(renal, Differential, by.x='Gene', by.y='Gene')
Curious <- curious[unique(curious$Gene),]
Curious <- Curious[complete.cases(Curious$DE_ctrl_adv),]
Curious <- Curious[!duplicated(Curious$Gene),]

KidneyGenesBothSets <- rbind(Expressions, Curious)
write.csv(KidneyGenesBothSets, 'KidneyDisease_Diabetes_Renal.csv',
          row.names=FALSE)
KidneyGenesBothSets[,1:3]
```


#### Expressions is the Diabetes data set of this study and the KidneyGenesBothSets is the combined data set of Diabetes and Renal disease genes that can be compared side by side. 

Using the Expressions data set, make the row names the Gene field while removing the statistically derived data from dplyer after writing to csv file.
```{r}
write.csv(Expressions, 'Diabetes_unique_genes_stats.csv', row.names=FALSE)
row.names(Expressions) <- Expressions$Gene
Expressions <- Expressions[,c(13:48)]
```

Do the same for the both data set of the two different studies's prominent genes
```{r}
row.names(KidneyGenesBothSets) <- KidneyGenesBothSets$Gene
BothStudies <- KidneyGenesBothSets[,c(13:48)]
```

Now BothStudies is the data set of genes prominent in both diabetes and renal disease gene expression data from kidney tissue tumors and healthy kidney tissue.

#### Transpose both data sets to see which one will perform better with machine learning on this diabetes kidney disease data, and add in a sample type for Advanced, Early, or healthy diabetes kidney tissue sample type.
```{r}
Expressions_t <- as.data.frame(t(Expressions))
type <- as.data.frame(c(rep('Advanced Diabetes',21), 
                        rep('Early Diabetes',6),
                        rep('Healthy', 9)))
colnames(type) <- 'TYPE'
Expressions_ML <- cbind(type, Expressions_t)
```

```{r}
BothStudies_t <- as.data.frame(t(BothStudies))
BothStudies_ML <- cbind(type, BothStudies_t)
```

Write both the machine learning ready data sets to csv
```{r}
write.csv(Expressions_ML, 'ML_ready_DiabetesGenes.csv', row.names=TRUE)
write.csv(BothStudies_ML, 'ML_ready_DiabetesAndRenalGenes.csv',
          row.names=TRUE)
```

The data set that will be used for Machine Learning will predict if the sample is renal disease or healthy. The samples will have to be randomized into 80% train and 20% test
```{r ,warning=FALSE,error=FALSE,message=FALSE}
library(caret)
library(randomForest)
#library(MASS)
#library(gbm)
#library(dplyr)
```

```{r cars}
set.seed(189678345)
```

#### The Diabetes genes are used first to predict type of sample as Advanced, Early, or healthy diabetes cases
```{r}
inTrain <- createDataPartition(y=Expressions_ML$TYPE, p=0.8, list=FALSE)

trainingSet <- Expressions_ML[inTrain,]
testingSet <- Expressions_ML[-inTrain,]

```



### KNN
```{r, message=FALSE, error=FALSE, warning=FALSE}
system.time(knnMod <- train(TYPE  ~ .,
                method='knn', preProcess=c('center','scale'),
                tuneLength=10, trControl=trainControl(method='cv'), data=trainingSet))

```

```{r}
plot(knnMod)
```

The predicted results with KNN
```{r}
predKNN <- predict(knnMod, testingSet)
predKNN
```

The actual values in the testing set
```{r}
testingSet$TYPE
```

The accuracy with K-Nearest Neighbors (KNN)
```{r}
accuracy <- (sum(predKNN==testingSet$TYPE)/length(predKNN))*100.00
acc <- paste(paste('The KNN accuracy is ',accuracy, sep=''),'%',sep='')
acc
```



### Random Forest
```{r, message=FALSE, error=FALSE, warning=FALSE}
system.time(rfMod <- train(TYPE ~., method='rf', data=(trainingSet), 
               trControl=trainControl(method='cv'), number=5))

```

```{r}
plot(rfMod)
```
The predicted Random Forest results and the actual results
```{r}
predRF <- predict(rfMod, testingSet)
predRF
testingSet$TYPE
```

The accuracy with Random Forest
```{r}
accuracy <- (sum(predRF==testingSet$TYPE)/length(predRF))*100.00
acc <- paste(paste('The Random Forest accuracy is ',accuracy, sep=''),'%',sep='')
acc
```


#### Lastly, the data set with genes prominent in this diabetes study and the study on renal disease on kidney tumor samples will be used to identify the same type of sample derived as Advanced Diabetes, Early Diabetes, or Healthy.
```{r}
inTrain <- createDataPartition(y=BothStudies_ML$TYPE, p=0.8, list=FALSE)

trainingSet <- BothStudies_ML[inTrain,]
testingSet <- BothStudies_ML[-inTrain,]

```



### KNN
```{r, message=FALSE, error=FALSE, warning=FALSE}
system.time(knnMod <- train(TYPE  ~ .,
                method='knn', preProcess=c('center','scale'),
                tuneLength=10, trControl=trainControl(method='cv'), data=trainingSet))

```

```{r}
plot(knnMod)
```

The predicted results with KNN
```{r}
predKNN <- predict(knnMod, testingSet)
predKNN
```

The actual values in the testing set
```{r}
testingSet$TYPE
```

The accuracy with K-Nearest Neighbors (KNN)
```{r}
accuracy <- (sum(predKNN==testingSet$TYPE)/length(predKNN))*100.00
acc <- paste(paste('The KNN accuracy is ',accuracy, sep=''),'%',sep='')
acc
```



### Random Forest
```{r, message=FALSE, error=FALSE, warning=FALSE}
system.time(rfMod <- train(TYPE ~., method='rf', data=(trainingSet), 
               trControl=trainControl(method='cv'), number=5))

```

```{r}
plot(rfMod)
```
The predicted Random Forest results and the actual results
```{r}
predRF <- predict(rfMod, testingSet)
predRF
testingSet$TYPE
```

The accuracy with Random Forest
```{r}
accuracy <- (sum(predRF==testingSet$TYPE)/length(predRF))*100.00
acc <- paste(paste('The Random Forest accuracy is ',accuracy, sep=''),'%',sep='')
acc
```


### Both gene sets are able to predict which type of sample the observation is from of either Advanced Diabetes, Early Diabetes, or Healthy kidney tissue samples.

Some text mining to build word clouds based on the NCBI gene summaries for these genes would be useful to show what genes are expressed more and see if their gene functions in the human body show up in the word clouds for up regulated (including increased fold change) and down regulated.


```{r}
both_ML <- KidneyGenesBothSets[,1:2]
diabetes_ML <- expressions[,1:2]
diabetes_ML <- diabetes_ML[unique(diabetes_ML$Gene),]

diabetes_down <- diabetes_ML[grep('decreased',diabetes_ML$Regulation),]
diabetes_up <- diabetes_ML[grep('increased',diabetes_ML$Regulation),]
both_down <- both_ML[grep('decreased', both_ML$Regulation),]
both_up <- both_ML[grep('increased', both_ML$Regulation),]

write.csv(diabetes_down, 'diabetes_down.csv', row.names=FALSE)
write.csv(diabetes_up, 'diabetes_up.csv', row.names=FALSE)
write.csv(both_down, 'both_down.csv', row.names=FALSE)
write.csv(both_up, 'both_up.csv', row.names=FALSE)
```

#### The gene functions were added to the diabetes_up and diabetes_down csv files from genecards.org
```{r}
# These files were modified in Excel to add the gene names and change the file #name so that when this script is ran again the files aren't erased or #replaced with the empty versions.

summ_up <- read.csv('diabetes_up1.csv', sep=',', header=TRUE,
                    na.strings=c('',' '))
summ_down <- read.csv('diabetes_down1.csv', sep=',', header=TRUE,
                      na.strings=c('',' '))

```

Remove the duplicate genes in summ_up and combine the genes that information was added to in the diabetes data with the 'both' versions and write out to replace NAs in Excel manually from genecards.org. The files will be renamed so that they aren't replaced with empty information when this script runs again.
```{r}
summ_up <- summ_up[!duplicated(summ_up$Gene),]
summ_down <- summ_down[!duplicated(summ_down$Gene),]

both_up <- read.csv('both_up.csv', sep=',', header=TRUE, 
                    na.strings=c('',' '))
both_down <- read.csv('both_down.csv', sep=',', header=TRUE, 
                      na.strings=c('',' '))

up_both <- merge(both_up, summ_up, by.x='Gene', by.y='Gene', 
                 all.x=TRUE)
down_both <- merge(both_down, summ_down, by.x='Gene', by.y='Gene',
                   all.x=TRUE)
up_both <- up_both[,c(1,2,4)]
down_both <- down_both[,c(1,2,4)]
colnames(up_both) <- c('Gene','Regulation', 'geneCardsFunction')
colnames(down_both) <- c('Gene','Regulation', 'geneCardsFunction')

write.csv(up_both, 'up_both.csv', row.names=FALSE)
write.csv(down_both, 'down_both.csv', row.names=FALSE)
```



The files for the both up and both down genes have been filled in manually and renamed to add a 1 at the end. Read those files in now, so that we will have the four data sets to work with to build our word clouds with the genecards/uniprot gene summaries. The Entrez summary was left out because it is too scientific to put in interpretable word clouds based on gene functions for up and down regulated genes in kidney disease samples related to diabetes and/or renal disease.
```{r}
both_up <- read.csv('up_both1.csv', sep=',', header=TRUE,
                    na.strings=c('',' '))
both_down <- read.csv('down_both1.csv', sep=',', header=TRUE,
                      na.strings=c('',' '))
```

```{r}
rm(down_both);rm(up_both)
```

#### Now we have the completed gene functions attached to the four separate data sets we will be using for our text mining.

Those being: 
summ_up and summ_down for the diabetic kidney disease targets

both_up and both_down for the combination of diabetes and renal disease kidney disease targets

Use lemmatization on the data sets with filled in gene summaries
```{r, error=FALSE, warning=FALSE, message=FALSE}
library(tm)
library(SnowballC)
library(wordcloud)
library(ggplot2)
library(textstem)


```


```{r}
lemma <- lemmatize_strings(summ_up$geneCardsFunction, dictionary=lexicon::hash_lemmas)

Lemma <- as.data.frame(lemma)
Lemma <- cbind(Lemma, summ_up)

colnames(Lemma)[1] <- 'lemmatized_summary'

write.csv(Lemma, 'Lemmatized_Diabetes_Up.csv', row.names=FALSE)

```

```{r}
lemma1 <- lemmatize_strings(summ_down$geneCardsFunction, dictionary=lexicon::hash_lemmas)

Lemma1 <- as.data.frame(lemma1)
Lemma1 <- cbind(Lemma1, summ_down)

colnames(Lemma1)[1] <- 'lemmatized_summary'

write.csv(Lemma1, 'Lemmatized_Diabetes_Down.csv', row.names=FALSE)

```

```{r}
lemma2 <- lemmatize_strings(both_up$geneCardsFunction, dictionary=lexicon::hash_lemmas)

Lemma2 <- as.data.frame(lemma2)
Lemma2 <- cbind(Lemma2, both_up)

colnames(Lemma2)[1] <- 'lemmatized_summary'

write.csv(Lemma2, 'Lemmatized_Diabetes_Renal_Up.csv', row.names=FALSE)

```

```{r}
lemma3 <- lemmatize_strings(both_down$geneCardsFunction, dictionary=lexicon::hash_lemmas)

Lemma3 <- as.data.frame(lemma3)
Lemma3 <- cbind(Lemma3, both_down)

colnames(Lemma3)[1] <- 'lemmatized_summary'

write.csv(Lemma3, 'Lemmatized_Diabetes_Renal_Down.csv', row.names=FALSE)

```


## Up regulated and increased fold change genes in diabetes mellitis (DM) samples


```{r, error=FALSE, message=FALSE, warning=FALSE}
dir.create('./DM-Up-Lemma')

ea <- as.character(Lemma$lemmatized_summary)
setwd('./DM-Up-Lemma')

for (j in 1:length(ea)){
  write(ea[j], paste(paste('Up',j, sep='.'), '.txt', sep=''))
}
setwd('../')

```


```{r}
DiabetesUp <- Corpus(DirSource("DM-Up-Lemma"))

DiabetesUp

DiabetesUp <- tm_map(DiabetesUp, removePunctuation)
DiabetesUp <- tm_map(DiabetesUp, removeNumbers)
DiabetesUp <- tm_map(DiabetesUp, tolower)
DiabetesUp <- tm_map(DiabetesUp, removeWords, stopwords("english"))
DiabetesUp <- tm_map(DiabetesUp, stripWhitespace)

dtmDiabetesUp <- DocumentTermMatrix(DiabetesUp)
dtmDiabetesUp

```


```{r}
freq <- colSums(as.matrix(dtmDiabetesUp))

FREQ <- data.frame(freq)
ord <- order(freq, decreasing=TRUE)

freq[head(ord, 25)]

```


### Up regulated genes
```{r, width=500, height=500}
wf <- data.frame(word=names(freq), freq=freq)
p <- ggplot(subset(wf, freq>10), aes(word, freq))
p <- p + geom_bar(stat= 'identity') 
p <- p + theme(axis.text.x=element_text(angle=90, hjust=1)) 
p

```

```{r}
wordcloud(names(freq), freq, min.freq=9,colors=brewer.pal(3,'Dark2'))

```

```{r}
wordcloud(names(freq), freq, max.words=30,colors=brewer.pal(6,'Dark2'))

```



## Down regulated genes in diabetes mellitis (DM) samples


```{r, error=FALSE, message=FALSE, warning=FALSE}
dir.create('./DM-Down-Lemma')

ea <- as.character(Lemma1$lemmatized_summary)
setwd('./DM-Down-Lemma')

for (j in 1:length(ea)){
  write(ea[j], paste(paste('Down',j, sep='.'), '.txt', sep=''))
}
setwd('../')

```


```{r}
DiabetesDown <- Corpus(DirSource("DM-Down-Lemma"))

DiabetesDown

DiabetesDown <- tm_map(DiabetesDown, removePunctuation)
DiabetesDown <- tm_map(DiabetesDown, removeNumbers)
DiabetesDown <- tm_map(DiabetesDown, tolower)
DiabetesDown <- tm_map(DiabetesDown, removeWords, stopwords("english"))
DiabetesDown <- tm_map(DiabetesDown, stripWhitespace)

dtmDiabetesDown <- DocumentTermMatrix(DiabetesDown)
dtmDiabetesDown

```


```{r}
freq <- colSums(as.matrix(dtmDiabetesDown))

FREQ <- data.frame(freq)
ord <- order(freq, decreasing=TRUE)

freq[head(ord, 25)]

```


### Down regulated genes
```{r, width=500, height=500}
wf <- data.frame(word=names(freq), freq=freq)
p <- ggplot(subset(wf, freq>10), aes(word, freq))
p <- p + geom_bar(stat= 'identity') 
p <- p + theme(axis.text.x=element_text(angle=90, hjust=1)) 
p

```

```{r}
wordcloud(names(freq), freq, min.freq=9,colors=brewer.pal(3,'Dark2'))

```

```{r}
wordcloud(names(freq), freq, max.words=30,colors=brewer.pal(6,'Dark2'))

```



### Both the Renal and Diabetes genes most expressed in Kidney tumor tissue gene expression data

```{r, error=FALSE, message=FALSE, warning=FALSE}
dir.create('./DM-RD-Up-Lemma')

ea <- as.character(Lemma2$lemmatized_summary)
setwd('./DM-RD-Up-Lemma')

for (j in 1:length(ea)){
  write(ea[j], paste(paste('BothUp',j, sep='.'), '.txt', sep=''))
}
setwd('../')

```


```{r}
BothUp <- Corpus(DirSource("DM-RD-Up-Lemma"))

BothUp

BothUp <- tm_map(BothUp, removePunctuation)
BothUp <- tm_map(BothUp, removeNumbers)
BothUp <- tm_map(BothUp, tolower)
BothUp <- tm_map(BothUp, removeWords, stopwords("english"))
BothUp <- tm_map(BothUp, stripWhitespace)

dtmBothUp <- DocumentTermMatrix(BothUp)
dtmBothUp

```


```{r}
freq <- colSums(as.matrix(dtmBothUp))

FREQ <- data.frame(freq)
ord <- order(freq, decreasing=TRUE)

freq[head(ord, 25)]

```


### Down regulated genes
```{r, width=500, height=500}
wf <- data.frame(word=names(freq), freq=freq)
p <- ggplot(subset(wf, freq>13), aes(word, freq))
p <- p + geom_bar(stat= 'identity') 
p <- p + theme(axis.text.x=element_text(angle=90, hjust=1)) 
p

```

```{r}
wordcloud(names(freq), freq, min.freq=14,colors=brewer.pal(3,'Dark2'))

```

```{r}
wordcloud(names(freq), freq, max.words=30,colors=brewer.pal(6,'Dark2'))

```



### Both the Renal and Diabetes genes Least expressed in Kidney tumor tissue gene expression data

```{r, error=FALSE, message=FALSE, warning=FALSE}
dir.create('./DM-RD-Down-Lemma')

ea <- as.character(Lemma2$lemmatized_summary)
setwd('./DM-RD-Down-Lemma')

for (j in 1:length(ea)){
  write(ea[j], paste(paste('BothDown',j, sep='.'), '.txt', sep=''))
}
setwd('../')

```


```{r}
BothDown <- Corpus(DirSource("DM-RD-Down-Lemma"))

BothDown

BothDown <- tm_map(BothDown, removePunctuation)
BothDown <- tm_map(BothDown, removeNumbers)
BothDown <- tm_map(BothDown, tolower)
BothDown <- tm_map(BothDown, removeWords, stopwords("english"))
BothDown <- tm_map(BothDown, stripWhitespace)

dtmBothDown <- DocumentTermMatrix(BothDown)
dtmBothDown

```


```{r}
freq <- colSums(as.matrix(dtmBothDown))

FREQ <- data.frame(freq)
ord <- order(freq, decreasing=TRUE)

freq[head(ord, 25)]

```


### Down regulated genes
```{r, width=500, height=500}
wf <- data.frame(word=names(freq), freq=freq)
p <- ggplot(subset(wf, freq>15), aes(word, freq))
p <- p + geom_bar(stat= 'identity') 
p <- p + theme(axis.text.x=element_text(angle=90, hjust=1)) 
p

```

```{r}
wordcloud(names(freq), freq, min.freq=15,colors=brewer.pal(3,'Dark2'))

```

```{r}
wordcloud(names(freq), freq, max.words=30,colors=brewer.pal(6,'Dark2'))

```

